README.md
# molgenis-magma

Magmascript evaluator in JavaScript.

## Development
`yarn dev` to run dev server during development.

`yarn test` to run tests.

`yarn build` to create new production build.

### How to commit
We use conventional commits to generate changelogs and release notes. Please check: https://www.conventionalcommits.org/

**Example**
```
git commit file.ext -m "fix(file.ext): fixes something"
```

### How to publish
Each time a PR is merged a release will be done to NPM. The CHANGELOG.md and GitHub release will be ammended. 

The version of the package is based upon convential commits. Check: http://commitizen.github.io/cz-cli/.


CHANGELOG.md
## [0.1.2](https://github.com/molgenis/molgenis-js-magma/compare/v0.1.1...v0.1.2) (2021-01-05)


### Bug Fixes

* force graceful-js version for gulp ([5157049](https://github.com/molgenis/molgenis-js-magma/commit/5157049))

## [0.1.1](https://github.com/molgenis/molgenis-js-magma/compare/v0.1.0...v0.1.1) (2019-06-07)


### Bug Fixes

* ci build ([63fcc62](https://github.com/molgenis/molgenis-js-magma/commit/63fcc62))

# [0.1.0](https://github.com/molgenis/molgenis-js-magma/compare/v0.0.9...v0.1.0) (2019-05-31)


### Bug Fixes

* **semantic-deps:** add plugins for auto release ([6b431e8](https://github.com/molgenis/molgenis-js-magma/commit/6b431e8))


### Features

* **autorelease:** added Jeninsfile to autorelease package ([5fa35ee](https://github.com/molgenis/molgenis-js-magma/commit/5fa35ee))

### 0.0.9
- Move Molgenis MagmaScript evaluator to npm module
- Let `MagmaScript.value()` method return wrapped value unchanged
- Rename `evalScript` to `evaluator`
- Let `evaluator` replace objects in the result with their id value
- Pick up bugfix and unit tests from forms library


test/setup/node.js
global.chai = require('chai');
global.sinon = require('sinon');
global.chai.use(require('sinon-chai'));
global.math = require('mathjs');
global.moment = require('moment');

require('babel-core/register');
require('./setup')();

/*
	Uncomment the following if your library uses features of the DOM,
	for example if writing a jQuery extension, and
	add 'simple-jsdom' to the `devDependencies` of your package.json

  Note that JSDom doesn't implement the entire DOM API. If you're using
  more advanced or experimental features, you may need to switch to
  PhantomJS. Setting that up is currently outside of the scope of this
  boilerplate.
*/
// import simpleJSDom from 'simple-jsdom';
// simpleJSDom.install();


test/setup/setup.js
module.exports = function(root) {
  root = root ? root : global;
  root.expect = root.chai.expect;

  beforeEach(() => {
    // Using these globally-available Sinon features is preferrable, as they're
    // automatically restored for you in the subsequent `afterEach`
    root.sandbox = root.sinon.sandbox.create();
    root.stub = root.sandbox.stub.bind(root.sandbox);
    root.spy = root.sandbox.spy.bind(root.sandbox);
    root.mock = root.sandbox.mock.bind(root.sandbox);
    root.useFakeTimers = root.sandbox.useFakeTimers.bind(root.sandbox);
    root.useFakeXMLHttpRequest = root.sandbox.useFakeXMLHttpRequest.bind(root.sandbox);
    root.useFakeServer = root.sandbox.useFakeServer.bind(root.sandbox);
  });

  afterEach(() => {
    delete root.stub;
    delete root.spy;
    delete root.math;
    delete root.moment;
    root.sandbox.restore();
  });
};


test/unit/evaluator.spec.js
import {evaluator} from '../../src/MagmaScript'
import moment from 'moment'

describe('evaluator', () => {
  describe('string comparison', () => {
    const expression = '$(\'myProp\').value() === "test"'
    it('should eval  "test === "test" as true ', () => {
      const entity = {myProp: 'test'}
      expect(evaluator(expression, entity)).to.equal(true)
    })
    it('should eval "test === "not test" as false ', () => {
      const entity = {myProp: 'not test'}
      expect(evaluator(expression, entity)).to.equal(false)
    })
  })

  describe('isValidJson', () => {
    const expression = '$(\'myProp\').isValidJson().value()'
    it('should return true in case of valid json ("{"foo": 5}")', () => {
      const entity = {myProp: '{"foo": 5}'}
      const result = evaluator(expression, entity)
      expect(result).to.equal(true)
    })
    it('should return false in case of invalid json ("{"foo: 5}"), missing quotes after foo', () => {
      const entity = {myProp: '{"foo: 5}'}
      const result = evaluator(expression, entity)
      expect(result).to.equal(false)
    })
  })

  describe('plus', () => {
    it('add property (3) to value (2) should equal 5', () => {
      const expression = '$(\'myProp\').plus(3).value()'
      const entity = {myProp: 2}
      const result = evaluator(expression, entity)
      expect(result).to.equal(5)
    })
    it('add property (2) to property (1) should equal 3', () => {
      const expression = '$(\'myProp1\').plus($(\'myProp2\')).value()'
      const entity = {myProp1: 2, myProp2: 1}
      const result = evaluator(expression, entity)
      expect(result).to.equal(3)
    })
  })

  describe('pow', () => {
    it('raising a property (3) to the power of 3 should equal 27', () => {
      const expression = '$(\'myProp\').pow(3).value()'
      const entity = {myProp: 3}
      const result = evaluator(expression, entity)
      expect(result).to.equal(27)
    })
  })

  describe('times', () => {
    it('property (2) times value (3) should equal 6', () => {
      const expression = '$(\'myProp\').times(3).value()'
      const entity = {myProp: 2}
      const result = evaluator(expression, entity)
      expect(result).to.equal(6)
    })
    it('property (2) times property (4) should equal 8', () => {
      const expression = '$(\'myProp1\').times($(\'myProp2\')).value()'
      const entity = {myProp1: 2, myProp2: 4}
      const result = evaluator(expression, entity)
      expect(result).to.equal(8)
    })
  })

  describe('div', () => {
    it('property (1) divided by value (1) should equal 0.5', () => {
      const expression = '$(\'myProp\').div(2).value()'
      const entity = {myProp: 1}
      const result = evaluator(expression, entity)
      expect(result).to.equal(0.5)
    })
    it('property (8) divided by property (2) should equal 4', () => {
      const expression = '$(\'myProp1\').div($(\'myProp2\')).value()'
      const entity = {myProp1: 8, myProp2: 2}
      const result = evaluator(expression, entity)
      expect(result).to.equal(4)
    })
  })

  describe('age', () => {
    const expression = '$("date").age().value()'

    it('should return undefined in case of empty property', () => {
      const entity = {date: null}
      const result = evaluator(expression, entity)
      expect(result).to.equal(undefined)
    })

    it('should return the age in years (4) if the property is a valid date', () => {
      const d = new Date()
      const year = d.getFullYear()
      const month = d.getMonth()
      const day = d.getDate()
      const testDate = new Date(year - 4, month, day)
      const entity = {date: testDate}
      const result = evaluator(expression, entity)
      expect(result).to.equal(4)
    })

    it('should return the age in years if the property is a valid date string', () => {
      const date = '2015-01-01'
      const entity = {date: date}
      const actual = evaluator(expression, entity)

      const expected = moment().diff(date, 'years')
      expect(actual).to.equal(expected)
    })
  })

  describe('map', () => {
    it('should map property value (1) to 2 given the mapping 1:2', () => {
      const expression = '$(\'myProp\').map({0:1, 1:2}, 4, 5).value()'
      const entity = {myProp: 1}
      const result = evaluator(expression, entity)
      expect(result).to.equal(2)
    })

    it('should map to a default value 4 if no mapping is defined for the property (3)', () => {
      const expression2 = '$(\'myProp\').map({0:1, 1:2}, 4, 5).value()'
      const entity2 = {myProp: 3}
      const result2 = evaluator(expression2, entity2)
      expect(result2).to.equal(4)
    })

    it('should map to the null value (5) is property is missing', () => {
      const expression2 = '$(\'myProp\').map({0:1, 1:2}, 4, 5).value()'
      const entity2 = {myProp: null}
      const result2 = evaluator(expression2, entity2)
      expect(result2).to.equal(5)
    })

    it('should map to the null value (5) is property is undefined', () => {
      const expression2 = '$(\'myProp\').map({0:1, 1:2}, 4, 5).value()'
      const entity2 = {myProp: null}
      const result2 = evaluator(expression2, entity2)
      expect(result2).to.equal(5)
    })
  })

  describe('group', () => {
    it('should transform value (101) into group (75+), given a set of groups [18, 35, 50, 75]', () => {
      const expression = '$(\'myProp\').group([18, 35, 50, 75]).value()'
      const entity = {myProp: 101}
      const result = evaluator(expression, entity)
      expect(result).to.equal('75+')
    })
  })

  describe('eq', () => {
    it('property value (null) compared to value null returns false', () => {
      const expression = '$(\'myProp\').eq().value()'
      const entity = {myProp: null}
      const result = evaluator(expression, entity)
      expect(result).to.equal(false)
    })

    it('property value (null) compared to non null value (3) returns false', () => {
      const expression = '$(\'myProp\').eq(3).value()'
      const entity = {myProp: null}
      const result = evaluator(expression, entity)
      expect(result).to.equal(false)
    })

    it('non null property value ("a") compared to non null value ("b") returns false', () => {
      const expression = '$(\'myProp\').eq("b").value()'
      const entity = {myProp: 'a'}
      const result = evaluator(expression, entity)
      expect(result).to.equal(false)
    })

    it('non null property value ("c") compared to non null value ("c") returns true', () => {
      const expression = '$(\'myProp\').eq("c").value()'
      const entity = {myProp: 'c'}
      const result = evaluator(expression, entity)
      expect(result).to.equal(true)
    })
  })

  describe('matches', () => {
    it('property value ("123abc456") matched against regex ("abc") should equal true', () => {
      var patt = new RegExp('abc')
      const expression = '$(\'myProp\').matches(' + patt + ').value()'
      const entity = {myProp: '123abc456'}
      const result = evaluator(expression, entity)
      expect(result).to.equal(true)
    })

    it('property value ("123abc456") matched against regex ("efg") should equal false', () => {
      var patt = new RegExp('efg')
      const expression = '$(\'myProp\').matches(' + patt + ').value()'
      const entity = {myProp: '123abc456'}
      const result = evaluator(expression, entity)
      expect(result).to.equal(false)
    })
  })

  describe('isNull', () => {
    it('should return true if property value is null', () => {
      const expression = '$(\'myProp\').isNull().value()'
      const entity = {myProp: null}
      const result = evaluator(expression, entity)
      expect(result).to.equal(true)
    })

    it('should return false if property value is not null', () => {
      const expression = '$(\'myProp\').isNull().value()'
      const entity = {myProp: 'not null'}
      const result = evaluator(expression, entity)
      expect(result).to.equal(false)
    })
  })

  describe('not', () => {
    it('should return the negation (true) of the property value (false)', () => {
      const expression = '$(\'myProp\').not().value()'
      const entity = {myProp: false}
      const result = evaluator(expression, entity)
      expect(result).to.equal(true)
    })
  })

  describe('or', () => {
    it('should return true if one property value is true and other is not', () => {
      const expression = '$(\'myProp\').or($(\'otherProp\')).value()'
      const entity = {myProp: true, otherProp: false}
      const result = evaluator(expression, entity)
      expect(result).to.equal(true)
    })

    it('should return true if one property value is false and the other is true', () => {
      const expression = '$(\'myProp\').or($(\'otherProp\')).value()'
      const entity = {myProp: false, otherProp: true}
      const result = evaluator(expression, entity)
      expect(result).to.equal(true)
    })

    it('should return true if both values are true', () => {
      const expression = '$(\'myProp\').or($(\'otherProp\')).value()'
      const entity = {myProp: true, otherProp: true}
      const result = evaluator(expression, entity)
      expect(result).to.equal(true)
    })

    it('should return false if both values are false', () => {
      const expression = '$(\'myProp\').or($(\'otherProp\')).value()'
      const entity = {myProp: false, otherProp: false}
      const result = evaluator(expression, entity)
      expect(result).to.equal(false)
    })
  })

  describe('and', () => {
    it('should return false if one property value is true and other is not', () => {
      const expression = '$(\'myProp\').and($(\'otherProp\')).value()'
      const entity = {myProp: true, otherProp: false}
      const result = evaluator(expression, entity)
      expect(result).to.equal(false)
    })

    it('should return false if one property value is false and the other is true', () => {
      const expression = '$(\'myProp\').and($(\'otherProp\')).value()'
      const entity = {myProp: false, otherProp: true}
      const result = evaluator(expression, entity)
      expect(result).to.equal(false)
    })

    it('should return true if both values are true', () => {
      const expression = '$(\'myProp\').and($(\'otherProp\')).value()'
      const entity = {myProp: true, otherProp: true}
      const result = evaluator(expression, entity)
      expect(result).to.equal(true)
    })

    it('should return false if both values are false', () => {
      const expression = '$(\'myProp\').and($(\'otherProp\')).value()'
      const entity = {myProp: false, otherProp: false}
      const result = evaluator(expression, entity)
      expect(result).to.equal(false)
    })
  })

  describe('gt', () => {
    it('should return true if the property value (5) is greater than the given value (4)', () => {
      const expression = '$(\'myProp\').gt(4).value()'
      const entity = {myProp: 5}
      const result = evaluator(expression, entity)
      expect(result).to.equal(true)
    })

    it('should return false if the property value (2) is less than the given value (9)', () => {
      const expression = '$(\'myProp\').gt(9).value()'
      const entity = {myProp: 2}
      const result = evaluator(expression, entity)
      expect(result).to.equal(false)
    })

    it('should return false if the property value is null', () => {
      const expression = '$(\'myProp\').gt(9).value()'
      const entity = {myProp: null}
      const result = evaluator(expression, entity)
      expect(result).to.equal(false)
    })
  })

  describe('lt', () => {
    it('should return false if the property value (5) is more then than the given value (4)', () => {
      const expression = '$(\'myProp\').lt(4).value()'
      const entity = {myProp: 5}
      const result = evaluator(expression, entity)
      expect(result).to.equal(false)
    })

    it('should return true if the property value (2) is less then than the given value (9)', () => {
      const expression = '$(\'myProp\').lt(9).value()'
      const entity = {myProp: 2}
      const result = evaluator(expression, entity)
      expect(result).to.equal(true)
    })

    it('should return false if the property value is null', () => {
      const expression = '$(\'myProp\').lt(9).value()'
      const entity = {myProp: null}
      const result = evaluator(expression, entity)
      expect(result).to.equal(false)
    })
  })

  describe('ge (greater then or equals)', () => {
    it('should return true if the property value (5) is greater than the given value (4)', () => {
      const expression = '$(\'myProp\').ge(4).value()'
      const entity = {myProp: 5}
      const result = evaluator(expression, entity)
      expect(result).to.equal(true)
    })

    it('should return false if the property value (2) is less than the given value (9)', () => {
      const expression = '$(\'myProp\').ge(9).value()'
      const entity = {myProp: 2}
      const result = evaluator(expression, entity)
      expect(result).to.equal(false)
    })

    it('should return true if the property value (2) is equal to the given value (2)', () => {
      const expression = '$(\'myProp\').ge(2).value()'
      const entity = {myProp: 2}
      const result = evaluator(expression, entity)
      expect(result).to.equal(true)
    })

    it('should return false if the property value is null', () => {
      const expression = '$(\'myProp\').ge(9).value()'
      const entity = {myProp: null}
      const result = evaluator(expression, entity)
      expect(result).to.equal(false)
    })
  })

  describe('le (less then or equals)', () => {
    it('should return false if the property value (5) is more then than the given value (4)', () => {
      const expression = '$(\'myProp\').le(4).value()'
      const entity = {myProp: 5}
      const result = evaluator(expression, entity)
      expect(result).to.equal(false)
    })

    it('should return true if the property value (2) is less then than the given value (9)', () => {
      const expression = '$(\'myProp\').le(9).value()'
      const entity = {myProp: 2}
      const result = evaluator(expression, entity)
      expect(result).to.equal(true)
    })

    it('should return true if the property value (2) is equal to the given value (2)', () => {
      const expression = '$(\'myProp\').ge(2).value()'
      const entity = {myProp: 2}
      const result = evaluator(expression, entity)
      expect(result).to.equal(true)
    })

    it('should return false if the property value is null', () => {
      const expression = '$(\'myProp\').le(9).value()'
      const entity = {myProp: null}
      const result = evaluator(expression, entity)
      expect(result).to.equal(false)
    })
  })

  describe('le (less then or equals)', () => {
    it('should return false if the property value (5) is more then than the given value (4)', () => {
      const expression = '$(\'myProp\').le(4).value()'
      const entity = {myProp: 5}
      const result = evaluator(expression, entity)
      expect(result).to.equal(false)
    })

    it('should return true if the property value (2) is less then than the given value (9)', () => {
      const expression = '$(\'myProp\').le(9).value()'
      const entity = {myProp: 2}
      const result = evaluator(expression, entity)
      expect(result).to.equal(true)
    })

    it('should return true if the property value (2) is equal to the given value (2)', () => {
      const expression = '$(\'myProp\').ge(2).value()'
      const entity = {myProp: 2}
      const result = evaluator(expression, entity)
      expect(result).to.equal(true)
    })

    it('should return false if the property value is null', () => {
      const expression = '$(\'myProp\').le(9).value()'
      const entity = {myProp: null}
      const result = evaluator(expression, entity)
      expect(result).to.equal(false)
    })
  })
})


src/magma-script-evaluator.js
import math from 'mathjs'
import moment from 'moment'

/**
 * Looks up the '_idValue' value in objects.
 * @param value object or array of objects
 * @returns {*} value with _idValue property replaced in objects
 */
export const toIdValue = (value) => {
  if(value === null) return null
  if(value === undefined) return undefined
  if(Array.isArray(value)) return value.map(toIdValue)
  if(typeof value === 'object' && value['_idValue'] !== undefined) return value._idValue
  return value
}

/**
 * Evaluates a script.
 *
 * @param script the script
 * @param entity object mapping attribute name key to attribute value
 * @returns the evaluated script result
 */
export function evaluator (script, entity) {
  // make functions available to the script in local scope
  // noinspection JSUnusedLocalSymbols
  const $ = MagmaScript.$.bind(entity) //NOSONAR
  // noinspection JSUnusedLocalSymbols
  const newValue = MagmaScript.newValue // NOSONAR
  // noinspection JSUnusedLocalSymbols
  const _isNull = MagmaScript._isNull // NOSONAR
  const value = eval(script)
  return toIdValue(value)
}

export function MagmaScript (val) {
  this.val = val
}

/**
 *
 * Gives you the value of the attribute specified between $('')
 * notation
 *
 * Example: $('Height').value() returns the values of the height
 * attribute
 *
 * If the value is an object, returns the idValue
 *
 * @memberof MagmaScript
 * @method value
 *
 */
MagmaScript.prototype.value = function () {
  return this.val
}

/**
 * Returns true if your value is valid JSON
 * Returns false if not
 */
MagmaScript.prototype.isValidJson = function () {
  try {
    JSON.parse(this.val)
    this.val = true
  } catch (e) {
    this.val = false
  }
  return this
}

/**
 * returns the result of the first value plus the second value
 *
 * @param value: the number you want to add to the current value
 */
MagmaScript.prototype.plus = function (value) {
  if (!MagmaScript._isNull(value)) {
    if (typeof value === 'object' || typeof value === 'function') {
      this.val = this.val + value.value()
    } else {
      this.val = this.val + value
    }
  }
  return this
}

/**
 * Gives you the exponent value of the attribute specified between
 * $('') notation
 *
 * Example: $('Height').pow(2).value() returns the result of
 * height_value ^ 2
 *
 * @param exp The number you use to execute the power function
 *
 * @memberof MagmaScript
 * @method pow
 */
MagmaScript.prototype.pow = function (exp) {
  if ((typeof denominator === 'object') && (typeof denominator.value === 'function')) {
    exp = exp.value()
  }
  this.val = Math.pow(this.val, exp)
  return this
}

/**
 * Gives you the multiplication value of the attribute specified between
 * $('') notation
 *
 * Example: $('Height').times(2).value() returns the result of
 * height_value * 2
 *
 * @param factor :
  *            The number you multiply by
 *
 * @memberof MagmaScript
 * @method div
 */
MagmaScript.prototype.times = function (factor) {
  if ((typeof factor === 'object') && (typeof factor.value === 'function')) {
    factor = factor.value()
  }
  this.val = (this.val * factor)
  return this
}

/**
 * Gives you the division value of the attribute specified between
 * $('') notation
 *
 * Example: $('Height').div(2).value() returns the result of
 * height_value / 2
 *
 * @param denominator The number you divide by
 *
 * @memberof MagmaScript
 * @method div
 */
MagmaScript.prototype.div = function (denominator) {
  if ((typeof denominator === 'object') && (typeof denominator.value === 'function')) {
    denominator = denominator.value()
  }
  this.val = (this.val / denominator)
  return this
}

/**
 * Returns the age based on the date of birth and the current year
 *
 * Example: $('Date_Of_Birth').age().value()
 *
 * @memberof MagmaScript
 * @method age
 */
MagmaScript.prototype.age = function () {
  if (MagmaScript._isNull(this.val)) {
    this.val = undefined
  } else {
    if (typeof this.val === 'string') {
      this.val = moment(this.val, 'YYYY-MM-DD', true).toDate()
    }
    this.val = Math.floor((Date.now() - this.val) / (365.2425 * 24 * 60 * 60 * 1000))
  }
  return this
}

/**
 * Maps categories to each other.
 *
 * Example: Dataset1 -> Male = 1, Female = 2 Dataset2 -> Male = 0,
 * Female = 1 $('Dataset2').map({0:1, 1:2}).value()
 *
 * @param categoryMapping  The mapping in JSON format to apply
 * @param defaultValue     a value to use for categories that are not mentioned in the categoryMappign
 * @param nullValue        a value to use for null instances
 *
 * @memberof MagmaScript
 * @method map
 */
MagmaScript.prototype.map = function (categoryMapping, defaultValue, nullValue) {
  if (typeof categoryMapping === 'function') {
    this.val = this.val.map(MagmaScript.newValue).map(categoryMapping)
  } else {
    const idValue = toIdValue(this.val)
    if (idValue in categoryMapping) {
      this.val = categoryMapping[idValue]
    } else {
      if (nullValue !== undefined && ((this.val === undefined) || (this.val === null))) {
        this.val = nullValue
      } else {
        this.val = defaultValue
      }
    }
  }
  return this
}

/**
 * Group values into defined ranges
 *
 * Example: age -> 19, 39, 50, 75
 * $('age').group({18, 35, 50, 75}).value() produces the following ranges which are left inclusive, (-∞, 18), [18, 35), [35, 50), [50, 75), [75, +∞)
 * the text representations are '-18','18-35','35-50','50-75','75+'
 *
 * @param arrayOfBounds a list of values that will be the bounds of the ranges.
 * @param arrayOfOutliers a list of outlier values that will be not be grouped and will be returned as is.
 * @param nullValue a value to use for null instances
 *
 * @memberof MagmaScript
 * @method group
 */
MagmaScript.prototype.group = function (arrayOfBounds, arrayOfOutliers, nullValue) {

  // Check if the the value is an outlier value
  if (arrayOfOutliers && arrayOfOutliers.length > 0) {
    for (let i = 0; i < arrayOfOutliers.length; i++) {
      if (this.val === arrayOfOutliers[i]) {
        return this
      }
    }
  }
  // find the ranges that the value fits into
  if (arrayOfBounds && arrayOfBounds.length > 0) {
    let originalValue = this.val
    if (originalValue < arrayOfBounds[0]) {
      this.val = '-' + arrayOfBounds[0]
    } else if (originalValue >= arrayOfBounds[arrayOfBounds.length - 1]) {
      this.val = arrayOfBounds[arrayOfBounds.length - 1] + '+'
    }
    if (arrayOfBounds.length > 1) {
      for (let i = 1; i < arrayOfBounds.length; i++) {
        const lowerBound = arrayOfBounds[i - 1]
        const upperBound = arrayOfBounds[i]
        //If lowerBound is bigger than upperBound, restore the original value and stop the function
        if (lowerBound > upperBound) {
          this.val = nullValue ? nullValue : null
          break
        }
        if (originalValue >= lowerBound && originalValue < upperBound) {
          this.val = lowerBound + '-' + upperBound
          break
        }
      }
    }
    return this
  }

  this.val = nullValue ? nullValue : null
  return this
}

/**
 * Compares two values and returns true or false
 *
 * Example: $('Height').eq(100).value()
 *
 * @param other the value you wish to compare with
 *
 * @memberof MagmaScript
 * @method eq
 */
MagmaScript.prototype.eq = function (other) {
  if (MagmaScript._isNull(this.val) && MagmaScript._isNull(other)) {
    this.val = false
  } else if (MagmaScript._isNull(this.val) && !MagmaScript._isNull(other)) {
    this.val = false
  } else {
    this.val = (this.value() === other)
  }
  return this
}

/**
 * Check if a value matches a regular expression
 *
 * Example: $('Username').matches(/^[a-z0-9_-]{6,18}$/).value()
 */
MagmaScript.prototype.matches = function (regex) {
  this.val = regex.test(this.val)
  return this
}

/**
 * Check if a value is null
 *
 * Example: $('Height').isNull().value()
 *
 * @memberof MagmaScript
 * @method isNull
 */
MagmaScript.prototype.isNull = function () {
  this.val = MagmaScript._isNull(this.val)
  return this
}

/**
 * Checks if a boolean is not
 *
 * Example: $('Has_Ears').not().value()
 *
 * @memberof MagmaScript
 * @method not
 */
MagmaScript.prototype.not = function () {
  this.val = !this.val
  return this
}

/**
 * Checks if something is one value or the other
 *
 * Example: $('male').or($('female')).value()
 *
 * @param other Another value
 *
 * @memberof MagmaScript
 * @method or
 */
MagmaScript.prototype.or = function (other) {
  this.val = (this.val || other.value())
  return this
}

/**
 * Checks if something is one value and the other
 *
 * Example: $('female').and($('pregnant')).value()
 *
 * @param other Another value
 *
 * @memberof MagmaScript
 * @method and
 */
MagmaScript.prototype.and = function (other) {
  this.val = (this.val && other.value())
  return this
}

/**
 * Returns true or false if Greater then the submitted value
 *
 * Example: $('Height').gt(100).value()
 *
 * @param value The value you compare with
 *
 * @memberof MagmaScript
 * @method gt
 */
MagmaScript.prototype.gt = function (value) {
  this.val = MagmaScript._isNull(this.val) ? false : (this.value() > value)
  return this
}

/**
 * Returns true or false if Less then the submitted value
 *
 * Example: $('Height').lt(100).value()
 *
 * @param value The value you compare with
 *
 * @memberof MagmaScript
 * @method lt
 */
MagmaScript.prototype.lt = function (value) {
  this.val = MagmaScript._isNull(this.val) ? false : (this.value() < value)
  return this
}
/**
 * Returns true or false if Greater or equal then the submitted
 * value
 *
 * Example: $('Height').ge(100).value()
 *
 * @param value The value you compare with
 *
 * @memberof MagmaScript
 * @method ge
 */
MagmaScript.prototype.ge = function (value) {
  this.val = MagmaScript._isNull(this.val) ? false : (this.value() >= value)
  return this
}
/**
 * Returns true or false if Less or equal then the submitted value
 *
 * Example: $('Height').le(100).value()
 *
 * @param value The value you compare with
 *
 * @memberof MagmaScript
 * @method le
 */
MagmaScript.prototype.le = function (value) {
  this.val = MagmaScript._isNull(this.val) ? false : (this.value() <= value)
  return this
}
/**
 * Sets the measurement unit of the current value to the specified
 * unit. Returns the current unit when no argument is supplied.
 *
 * @memberof MagmaScript
 * @method unit
 */
MagmaScript.prototype.unit = function (newUnit) {
  if (!newUnit) {
    return this.unit
  }
  this.unit = newUnit
  return this
}

/**
 * Measurement unit conversion: converts the current value into a
 * different measurement unit.
 *
 * @memberof MagmaScript
 * @method toUnit
 */
MagmaScript.prototype.toUnit = function (targetUnit) {
  const unit = math.unit(this.val, this.unit)
  this.val = unit.toNumber(targetUnit)
  return this
}

/**
 * Stores the computed MagmaScript value after applying on of the mathematical
 * functions listed below
 *
 * In case of deep referencing attribute e.g. xref.xref2.label, we loop through these attributes and only return
 * new MagmaScript(label)
 *
 * @version 1.0
 * @namespace MagmaScript
 */
MagmaScript.$ = function (attr) {
  const attributes = attr.split('\.')
  let result = this // we expect the $ function be bound to the entity we're evaluating

  for (let i = 0; i < attributes.length && result !== null; i++) {
    result = result[attributes[i]]
  }
  return new MagmaScript(result)
}

MagmaScript._isNull = function (value) {
  if (value === null || value === undefined)
    return true
  return (typeof value === 'string') && (value.length === 0)
}

MagmaScript.newValue = function (value) {
  return new MagmaScript(value)
}

export default MagmaScript
